File: sources/dart/13_matrix_mul.dart
// task 13 matrix_mul — expected output: 599995000
// build: dart compile exe -o prog 13_matrix_mul.dart (aot; jit has no build step)    run: ./prog (aot) | dart 13_matrix_mul.dart (jit)

import 'dart:typed_data';

void main() {
  const int n = 500;
  final Int32List a = Int32List(n * n);
  final Int32List b = Int32List(n * n);

  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      a[i * n + j] = (i + j) % 7;
    }
  }
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      b[i * n + j] = (i * j) % 5;
    }
  }

  final Int64List c = Int64List(n * n);
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      int sum = 0;
      for (int k = 0; k < n; k++) {
        sum += a[i * n + k] * b[k * n + j];
      }
      c[i * n + j] = sum;
    }
  }

  int total = 0;
  for (int i = 0; i < n * n; i++) {
    total += c[i];
  }
  print(total);
}

File: sources/zig/13_matrix_mul.zig
// task 13 matrix_mul — expected output: 599995000
// build: zig build-exe -O ReleaseFast 13_matrix_mul.zig -femit-bin=prog    run: ./prog
// deviation: Zig 0.16 removed std.posix.write, so the line goes out through std.Io.File.stdout().

const std = @import("std");

const n: usize = 500;

pub fn main(init: std.process.Init) !void {
    const io = init.io;
    const allocator = std.heap.page_allocator;

    const a = try allocator.alloc(i64, n * n);
    defer allocator.free(a);
    const b = try allocator.alloc(i64, n * n);
    defer allocator.free(b);
    const c = try allocator.alloc(i64, n * n);
    defer allocator.free(c);

    var i: usize = 0;
    while (i < n) : (i += 1) {
        var j: usize = 0;
        while (j < n) : (j += 1) {
            a[i * n + j] = @intCast((i + j) % 7);
            b[i * n + j] = @intCast((i * j) % 5);
        }
    }

    i = 0;
    while (i < n) : (i += 1) {
        var j: usize = 0;
        while (j < n) : (j += 1) {
            var sum: i64 = 0;
            var k: usize = 0;
            while (k < n) : (k += 1) {
                sum += a[i * n + k] * b[k * n + j];
            }
            c[i * n + j] = sum;
        }
    }

    var total: i64 = 0;
    var index: usize = 0;
    while (index < n * n) : (index += 1) {
        total += c[index];
    }

    var line_buf: [32]u8 = undefined;
    const line = try std.fmt.bufPrint(&line_buf, "{d}\n", .{total});
    try std.Io.File.stdout().writeStreamingAll(io, line);
}
